interpolation.go

```diff
@@ -60,3 +62,133 @@ func Interpolate(values []float64, size int) []float64 {
 	valuesNew[size-1] = values[oldSize-1]
 	return valuesNew
 }
+
+/**
+ * InterP1 simulate the interp1 method in matlab.
+ * It returns interpolated values of a 1-D function at specific query points using linear interpolation.
+ *
+ * This is ported from the js code which is ported from the matlab code:
+ * https://www.npmjs.com/package/interp1?activeTab=code
+ *
+ * @param  xList        Array of independent sample points.
+ * @param  vList        Array of dependent values v(x) with length equal to xList.
+ * @param  qeuryList    Array of coordinates of the query points.
+ * @return              Interpolated values with length equal to queryList.
+ */
+func InterP1(xList, vList []float64, queryList []float64) ([]float64, error) {
+	if len(xList) != len(vList) {
+		return nil, fmt.Errorf("arrays of sample points xList and corresponding values vList have to have equal length.: %d vs. %d\n", len(xList), len(vList))
+	}
+	if len(xList) == 0 {
+		return nil, fmt.Errorf("arrays of sample points xList and corresponding values vList have to have length > 0.")
+	}
+
+	type Zip struct {
+		X, V float64
+	}
+	// Combine x and v
+	p := []Zip{}
+	for i := range xList {
+		p = append(p, Zip{xList[i], vList[i]})
+	}
+
+	// Sort asc
+	sort.Slice(p, func(i, j int) bool {
+		return p[i].X < p[j].X
+	})
+
+	// Check for double x values
+	for i, v := range p[1:] {
+		if p[i].X == v.X {
```

Comment:
I don't get this. p[i] and v are the same variable. Wouldn't they always have the same X value?

Reply:
`p[i]` should be the value previous to `v`. But I think this logic is actually not working as expected.

@wmohrman I saw the original js code is like this

```
    /* Combine x and v arrays. */
    var zipped = xs.map(function (x, index) { return [x, vs[index]]; });
    /* Sort points by independent variabel in ascending order. */
    zipped.sort(function (a, b) {
        var diff = a[0] - b[0];
        /* Check if some x value occurs twice. */
        if (diff === 0) {
            throw new Error('Two sample points have equal value ' + a[0] + '. This is not allowed.');
        }
        return diff;
    });
```

It looks different from the logic we have here. The js code doesn't assume that the two sample points with equal values are always adjacent. It checks for equality between the values of the independent sample points (a[0] and b[0]) for any pair of elements being compared during the sorting. In the go code, we only check the two adjacent sample points. Do you know what's the logic supposed to be in Matlab?

Reply:
That js code looks wonky. It isn't guaranteed that the sort function compares every value against every other value so that will never check all possibly combinations. If you check every adjacent value after it is sorted though, you will guarantee that all values are different.

I'd say keep the go code but just change to 
```go
if p[i-1].X == v.X
```

Reply:
`p[i].X` is the value previous to `v.X`, because the range is on `p[1:]`, so ya the golang code is unnecessarily smart.

Maybe we should do something like

```
for i := range p {
    if i == 0 {
        continue
    }
    if p[i-1].X == p[i].X {
        xxxxxx
    }
}
```

Reply:
[f4caab1](https://github.com/stryd/dsp/pull/3/commits/f4caab12b447be29080812ccdaba6335ad98f35e) refactors everything